From: Hannes Reinecke <hare@suse.de>
Subject: mapped_base for S/390

S/390 needs special care for the mapped_base patch as s390x can operate
in either 31 or 64 bit mode. When running 64 bit programs any mapped_base
quite pointless.

--- linux-2.6.11/include/asm-s390/processor.h.orig	2005-06-16 15:04:13.357425944 +0200
+++ linux-2.6.11/include/asm-s390/processor.h	2005-06-16 15:27:49.797425944 +0200
@@ -59,23 +59,31 @@ extern struct task_struct *last_task_use
 /*
  * User space process size: 2GB for 31 bit, 4TB for 64 bit.
  */
+#define TASK31_SIZE		(0x80000000UL)
+#define TASK64_SIZE		(0x40000000000UL)
+
 #ifndef __s390x__
 
-# define TASK_SIZE		(0x80000000UL)
-# define TASK_UNMAPPED_BASE	(TASK_SIZE / 2)
-# define DEFAULT_TASK_SIZE	(0x80000000UL)
+# define TASK_SIZE		TASK31_SIZE
+# define TASK_UNMAPPED_BASE	(current->map_base)
+# define __TASK_UNMAPPED_BASE	(TASK_SIZE / 2)
+# define DEFAULT_TASK_SIZE	TASK31_SIZE
 
 #else /* __s390x__ */
 
 # define TASK_SIZE		(test_thread_flag(TIF_31BIT) ? \
-					(0x80000000UL) : (0x40000000000UL))
-# define TASK_UNMAPPED_BASE	(TASK_SIZE / 2)
-# define DEFAULT_TASK_SIZE	(0x40000000000UL)
+					TASK31_SIZE : TASK64_SIZE)
+# define TASK_UNMAPPED_BASE	(test_thread_flag(TIF_31BIT) ? \
+					(current->map_base) : (TASK64_SIZE / 2))
+# define __TASK_UNMAPPED_BASE	(TASK_SIZE / 2)
+# define DEFAULT_TASK_SIZE	TASK64_SIZE
 
 #endif /* __s390x__ */
 
 #define HAVE_ARCH_PICK_MMAP_LAYOUT
 
+#define MM_VM_SIZE(mm)		DEFAULT_TASK_SIZE
+
 typedef struct {
         __u32 ar4;
 } mm_segment_t;
